### PR TITLE
TS-28611 Implemented passing access key securely via stdin or environ…

### PR DESCRIPTION
**Pull request by j-christl**
…ment variable; Reformatted code

Addresses issue TS-28611

- [x] Changes are tested adequately
- [x] Teamscale documentation updated in case of relevant user-visible changes
- [x] CHANGELOG.md updated
- [ ] Present new features in [N&N](https://wiki.cqse.eu/pages/viewpage.action?pageId=689566)

Please respect the vote of the [Teamscale bot](https://demo.teamscale.com) or flag irrelevant findings as tolerated or
false positives. If you feel that the Teamscale config or architecture file needs adjustment, please state so in a
comment and discuss this with your reviewer.

### CODE REVIEW

File: .github/workflows/actions.yml [hunk 1]
```diff
@@ -49,7 +49,7 @@ jobs:
           native-image: true
       - name: Build
         env:
-          ACCESS_KEY: ${{ secrets.TEAMSCALE_ACCESS_KEY }}
+          TEAMSCALE_ACCESS_KEY: ${{ secrets.TEAMSCALE_ACCESS_KEY }}
```
Comment (saileru) [inline .github/workflows/actions.yml:52]: As we now support passing the access key via environment variable directly, we can also use this for the tests. Previously, we only read the access key from the environment variable to put it into the command line option for the tests. I adjusted the tests accordingly and also made the naming of the enviromnent variables consistent here for the access key which is read from the Github secret.

File: src/main/java/com/teamscale/upload/CommandLine.java [hunk 2]
```diff
@@ -110,6 +114,30 @@ private CommandLine(Namespace namespace) {
...
+		} else if (accessKeyViaOption != null) {
+			System.out.print("Reading access key from standard input: ");
+			Scanner inputScanner = new Scanner(System.in);
+			String accessKeyViaStdin = inputScanner.nextLine();
+			inputScanner.close();
+			System.out.println("\nRead access key");
+			return accessKeyViaStdin;
+		}
+		return System.getenv("TEAMSCALE_ACCESSKEY"); // may be null, but is validated later
 	}
```
[the remark this answers is not in the document]
Comment (saileru) [reply src/main/java/com/teamscale/upload/CommandLine.java:156]: I did this already in my rework. Please have a look if this makes sense 🙂

**Review by saileru — CHANGES_REQUESTED** [review]
@yoshc Thanks for implementing this. I thought it would be valuable to have a automatic tests for the new functionality, so I spent some time and added a few tests in a new commit. In the process I also did some minor rework. It would be great if you could review my changes 🙂

File: src/main/java/com/teamscale/upload/CommandLine.java [hunk 3]
```diff
@@ -30,7 +31,9 @@
 	 */
 	public final String username;
 	/**
-	 * The Teamscale access key.
+	 * Teamscale access key used for authentication. Either obtained via
+	 * command-line option, via stdin or via the environment variable
+	 * $TEAMSCALE_ACCESSKEY.
```
[the remark this answers is not in the document]
Comment (saileru) [reply src/main/java/com/teamscale/upload/CommandLine.java]: Did that already 🙂

File: src/test/java/com/teamscale/upload/NativeImageIT.java [hunk 4]
```diff
@@ -362,8 +436,12 @@ private Arguments withInput(String input) {
...
 
 		private String[] toStringArray() {
-			List<String> arguments = new ArrayList<>(Arrays.asList("-s", url, "-u", user, "-a", accessKey, "-f", format,
-					"-p", project, "-t", partition));
+			List<String> arguments = new ArrayList<>(
+					Arrays.asList("-s", url, "-u", user, "-f", format, "-p", project, "-t", partition));
+			if (accessKey != null) {
+				arguments.add("-a");
+				arguments.add(accessKey);
+			}
```
Comment (saileru) [inline src/test/java/com/teamscale/upload/NativeImageIT.java:444]: I would tend to tolerate the long-method finding here. I do not see a way to improve this by extracting methods...

File: src/main/java/com/teamscale/upload/CommandLine.java [hunk 5]
```diff
@@ -315,9 +331,11 @@ private void validateKeystoreSettings(ArgumentParser parser) throws ArgumentPars
 
 	private void validateAccessKey(ArgumentParser parser) throws ArgumentParserException {
 		if (accessKey == null) {
-			throw new ArgumentParserException("You did not specify a Teamscale access key. You can either specify"
-					+ "it via --accesskey <access key>, via setting the environment variable $TEAMSCALE_ACCESSKEY or via stdin"
-					+ " using '--accesskey -'.", parser);
+			throw new ArgumentParserException(
+					"You did not specify a Teamscale access key. You can either specify"
+							+ "it via --accesskey <access key>, via setting the environment variable $"
+							+ TEAMSCALE_ACCESS_KEY_ENVIRONMENT_VARIABLE + " or via stdin" + " using '--accesskey -'.",
```
Comment (yoshc) [inline src/main/java/com/teamscale/upload/CommandLine.java:337]: I think the last string concatenation got mixed up by the change and can be removed. Will do this in my rework

File: src/test/java/com/teamscale/upload/NativeImageIT.java [hunk 6]
```diff
@@ -290,6 +346,7 @@ private String extractNormalizedMessage(TeamscaleMockServer.Session session) {
 		private String commit = null;
 		private String additionalMessageLine = null;
 		private boolean stackTrace = false;
+		private String stdin = null;
```
[the remark this answers is not in the document]
Comment (j-christl) [reply src/test/java/com/teamscale/upload/NativeImageIT.java:349]: Refactored

File: src/test/java/com/teamscale/upload/NativeImageIT.java [hunk 7]
```diff
@@ -51,9 +55,8 @@ public void incorrectUrl() {
 		assertSoftlyThat(softly -> {
 			softly.assertThat(result.exitCode).isNotZero();
 			// the command line library we use adjusts the word spacing based on the
-			// terminal width
-			// so on different machines the output my contain a different number of spaces
-			// this behaviour can unfortunately not be disabled
+			// terminal width so on different machines the output may contain a different
+			// number of spaces this behaviour can unfortunately not be disabled
```
[the remark this answers is not in the document]
Comment (j-christl) [reply src/test/java/com/teamscale/upload/NativeImageIT.java:59]: Added dot

File: src/main/java/com/teamscale/upload/autodetect_revision/ProcessUtils.java [hunk 8]
```diff
@@ -21,11 +26,27 @@
...
+				input = new ByteArrayInputStream(accessKey.getBytes(StandardCharsets.ISO_8859_1));
+			} catch (IOException e) {
+				System.err.println("Could not read access key from file `" + stdinFile + ".");
+				return new ProcessResult(-1, "", e);
+			}
+		}
+
 		DefaultExecutor executor = new DefaultExecutor();
-		CaptureStreamHandler handler = new CaptureStreamHandler();
+		CaptureStreamHandler handler = new CaptureStreamHandler(input);
```
[the remark this answers is not in the document]
Comment (saileru) [reply src/main/java/com/teamscale/upload/autodetect_revision/ProcessUtils.java:54]: I had a look at the source code: https://commons.apache.org/proper/commons-exec/apidocs/src-html/org/apache/commons/exec/PumpStreamHandler.html#line.76
```
075    /**
076     * Construct a new <CODE>PumpStreamHandler</CODE>.
077     *
078     * @param out the output <CODE>OutputStream</CODE>.
079     * @param err the error <CODE>OutputStream</CODE>.
080     */
081    public PumpStreamHandler(final OutputStream out, final OutputStream err) {
082        this(out, err, null);
083    }
```
There, they also pass null as `input` in the two-argument constructor. So, I assume that `null` is a valid value here and simply means "no input stream".

**Review by saileru — APPROVED** [review]
Nice, thanks 🙂

**Pull request merged**